Product quantization (#114)
qdrant/docs: qdrant/v1.2.x/quantization.md

## Product Quantization

> *Available in Qdrant since v1.2.0*

Reviewer: I believe we use this everywhere else:

```suggestion
*Available since v1.2.0*
```

@davidmyriel suggested to use `Available as of v1.2.0`, but I think we should change that everywhere in a separate PR.

Author: I propose to do it once in a separate PR

Reviewer 2: ill take care of all of these little details in a doc overhaul

> Product quantization is a method of compressing vectors to minimize their memory usage by dividing them into 

Reviewer: Instead of directly diving into the technical stuff here, I think it would be useful to add a single sentence that broadly describes what this is.

For scalar quantization we have this:

> Scalar quantization, in the context of vector search engines, is a compression technique that compresses vectors by reducing the number of bits used to represent each vector component.

Maybe we can add something like this:

> Product quantization, in the context of vector search engines, is a compression technique that compresses vectors by reducing the number of dimensions by grouping a number of elements in buckets, to reduce the conditionality when searching.

(this is definitely wrong, so please correct it :) )

> chunks and quantizing each segment individually.

Reviewer 3: What's the relation between chunks and segments?

Each chunk is approximated by a centroid index that represents the original vector component.
The positions of the centroids are determined through the utilization of a clustering algorithm such as k-means.

> For now, Qdrant uses only 256 centroids, so each centroid index can be represented by a single byte.

Reviewer: > so each centroid index can be represented by a single byte.

Smart! I wasn't aware of that.

Product quantization can compress by a more prominent factor than a scalar one.

> But there are some tradeoffs. Product quantization distance calculation is not SIMD-friendly, so it is slower than scalar quantization.

Reviewer: ```suggestion
But there are some tradeoffs. Product quantization distance calculations are not SIMD-friendly, so it is slower than scalar quantization.
```

### Setting up Product Quantization

To enable product quantization, you need to specify the quantization parameters in the `quantization_config` section of the collection configuration.

```http
PUT /collections/{collection_name}

{
    "vectors": {
      "size": 768,
      "distance": "Cosine"
    },
    "quantization_config": {
        "product": {
            "compression": "x16",
            "always_ram": true
        }
    }
}
```

```python
from qdrant_client import QdrantClient
from qdrant_client.http import models

client = QdrantClient("localhost", port=6333)

client.recreate_collection(
    collection_name="{collection_name}",
    vectors_config=models.VectorParams(size=768, distance=models.Distance.COSINE),
    quantization_config=models.ProductQuantization(
        scalar=models.ProductQuantizationConfig(
            compression=models.CompressionRatio.X16,
            always_ram=True,
        ),
    ),
)
```

There are 2 parameters that you can specify in the `quantization_config` section:

`compression` - compression ratio.
Compression ratio represents the size of the quantized vector in bytes divided by the size of the original vector in bytes.

> In this xase, the quantized vector will be 16 times smaller than the original vector.

Reviewer 3: typo
```suggestion
In this case, the quantized vector will be 16 times smaller than the original vector.
```

{
    "vectors": {
      "size": 768,
      "distance": "Cosine"
    },
    "quantization_config": {
        "product": {
            "compression": "x16",
            "always_ram": true
        }
    }
}
```

```python
from qdrant_client import QdrantClient
from qdrant_client.http import models

client = QdrantClient("localhost", port=6333)

client.recreate_collection(
    collection_name="{collection_name}",
    vectors_config=models.VectorParams(size=768, distance=models.Distance.COSINE),
    quantization_config=models.ProductQuantization(
        scalar=models.ProductQuantizationConfig(
            compression=models.CompressionRatio.X16,
            always_ram=True,
        ),
    ),
)
```

There are 2 parameters that you can specify in the `quantization_config` section:

`compression` - compression ratio.
Compression ratio represents the size of the quantized vector in bytes divided by the size of the original vector in bytes.
In this xase, the quantized vector will be 16 times smaller than the original vector.

`always_ram` - whether to keep quantized vectors always cached in RAM or not. By default, quantized vectors are loaded in the same way as the original vectors.

> However, in some setups you might want to keep quantized vectors in RAM to speed up the search process.
> 
> In this case, you can set `always_ram` to `true` to store quantized vectors in RAM.

Reviewer: ```suggestion
However, in some setups you might want to keep quantized vectors in RAM to speed up the search process. Then set `always_ram` to `true`.
```
